csrc/id_graphs.h

```diff
@@ -0,0 +1,465 @@
+  // Note, any iter domains used during something like loop or concrete id
+  // resolution could actually have multiple Expr* uses, and uses on disjoint id
+  // sets should be used, not this.
+  //
+  // TODO: Refactor or remove?
+  Expr* idUse(IterDomain* id) const;
+  Expr* idDef(IterDomain* id) const;
+
+  // TODO: Seems a bit unfortunate that this isn't IterDomain local information.
+  const std::unordered_set<IterDomain*>& viewRfactorIds() const {
+    return view_rfactor_ids_;
+  }
+
+  // Returns if a self mapping was detected that would invalidate assumptions of
+  // the overall lowering system.
+  //
+  // TODO: Can we make this more of an alias analysis?
+  // Ref: https://github.com/csarofeen/pytorch/pull/1954#discussion_r961940498
+  bool hasSelfMapping() const {
+    return self_mapping_info_.has_value();
+  }
+
+  // Update the LOOP ID disjoint sets with resolved computeWith
+  void updateComputeWith(TensorView* compute_with_tv);
+
+  std::string toString() const;
+
+  // Replay Expr but with the inputs provided. IterDomainGraphss will be updated
+  // for all maps that have entries, adding the output iter domains of the
+  // replayed expression and adding potential mappings through the expression.
+  Expr* addReplayAs(const std::vector<IterDomain*>& new_inputs, Expr* expr);
+
+ protected:
+  // Sometimes fusion inputs or outputs are disconnected from expressions, in
+  // those cases we still may want to send in some additional tensor views from
+  // the Fusion that don't have expressions associated with them.
+  void build(
+      const std::vector<Expr*>& exprs,
+      const std::vector<TensorView*>& additional_tvs);
+
+  // ======= START Iteration domain build process in order called =======
+
+  // Fills id_uses_ and id_definitions_ for all IterDomains active in the
+  // fusion.
+  void buildIterDomainDefinitionsAndUses(
+      const std::vector<TensorView*>& all_tvs);
+
+  // Iterates over all IterDomains in id_definitions_ and calls initializeID on
+  // a new IdGraph and returns it.
+  IdGraph initializeIdGraph();
+
+  // Fills disjoint_ids_[IdMappingMode::EXACT] for relationships between inputs
+  // and first output of expr
+  void buildExactMap(const std::vector<Expr*>& exprs);
+
+  // Fills disjoint_ids_[IdMappingMode::ALMOSTEXACT]. Initialize AlmostExact as
+  // Exact entries, then map anything that's either merged with a size-1 or
+  // split by a size-1 dimension.
+  void buildAlmostExactMap();
+
+  // Fills disjoint_ids_[IdMappingMode::PERMISSIVE]. Initialize PermissiveMap as
+  // AlmostExact entries, then map through broadcasts
+  void buildPermissiveMap(const std::vector<Expr*>& exprs);
+
+  //! Run through disjoint sets in the LOOP map, make sure there's only one
+  //! non-serial parallel type in each disjoint set, set the parallel type of
+  //! all IterDomains in the disjoint set to that PType.
+  void validateAndPropagatePType() const;
+
+  void buildLoopPromotionMap(const std::vector<Expr*>& exprs);
```

Comment:
Basic question, but I see "promotion" mentioned many times here. In this context what does it mean to promote an ID?

Reply:
Promotion is the concept that an iteration domain that a TensorView has in its root->domain might not really be what's required for the generated kernel to index into that TensorView. Promotion is for example:
    Producer has a broadcast merged with an iteration domain
    Consumer has a (mapped to the producer) iteration domain merged with an iteration domain
    Based on other transformations the producer might have to "promote" it's broadcast domain to the iteration domain in it's consumer
    If there's a producer of that producer, then we still might need the "broadcast promotion" but there isn't a broadcast that maps in that producer's producer

Reply:
Revisiting this after some thought and going through some of the machinery. Just to try and reiterate the example above:
```c++
t0[(b0*i1), i2] // producer
t1[(i3 * i4)] = f(t0)  // consumer, through expression f
```
`i3` is defined by an expression in the variables `i1, i2`.

We may need to alter `b0`  to match `i3` if `i4` matches `i1`.

/This might cascade, since `(b0*i1)` might match another merged broadcast in _its_ producer.

End goal of promotion:  map each leaf IterDomain in each tensorview (importantly including bcast domains and transforms thereof) to an `Iteration` IterDomain that is written as a transform of that TVs root_domain, so that we can index into it.